src/hash.cr

```diff
@@ -1047,91 +1640,65 @@ class Hash(K, V)
   # {"foo" => "bar", "baz" => "bar"}.invert # => {"bar" => "baz"}
   # ```
   def invert
-    hash = Hash(V, K).new(initial_capacity: @buckets_size)
+    hash = Hash(V, K).new(initial_capacity: @size)
     self.each do |k, v|
       hash[v] = k
     end
     hash
   end
 
-  protected def find_entry(key)
-    return nil if empty?
-
-    index = bucket_index key
-    entry = @buckets[index]
-    find_entry_in_bucket entry, key
-  end
+  struct Entry(K, V)
+    getter key, value, hash
 
-  private def insert_in_bucket(index, key, value)
-    entry = @buckets[index]
-    if entry
-      while entry
-        if entry.key == key
-          entry.value = value
-          return nil
-        end
-        if entry.next
-          entry = entry.next
-        else
-          return entry.next = Entry(K, V).new(key, value)
-        end
-      end
-    else
-      return @buckets[index] = Entry(K, V).new(key, value)
+    def initialize(@hash : UInt32, @key : K, @value : V)
     end
-  end
 
-  private def find_entry_in_bucket(entry, key)
-    while entry
-      if entry.key == key
-        return entry
-      end
-      entry = entry.next
+    def self.deleted
+      key = uninitialized K
+      value = uninitialized V
+      new(0_u32, key, value)
     end
-    nil
-  end
-
-  private def bucket_index(key)
-    key.hash.remainder(@buckets_size).to_i
-  end
 
-  private def calculate_new_size(size)
-    new_size = 8
-    HASH_PRIMES.each do |hash_size|
-      return hash_size if new_size > size
-      new_size <<= 1
+    def deleted?
+      @hash == 0_u32
     end
-    raise "Hash table too big"
-  end
-
-  private class Entry(K, V)
-    getter key : K
-    property value : V
-
-    # Next in the linked list of each bucket
-    property next : self?
 
-    # Next in the ordered sense of hash
-    property fore : self?
-
-    # Previous in the ordered sense of hash
-    property back : self?
-
-    def initialize(@key : K, @value : V)
+    def matches?(hash, key)
+      # Tiny optimization: for these primitive types it's faster to just
+      # compare the key instead of comparing the hash and the key.
+      # We still have to skip hashes with value 0 (means deleted).
+      {% if K == Bool ||
+              K == Char ||
+              K == Symbol ||
+              K < Int::Primitive ||
+              K < Float::Primitive ||
+              K < Enum %}
+        @key == key && @hash != 0_u32
+      {% else %}
+        @hash == hash && @key == key
+      {% end %}
     end
   end
 
   private module BaseIterator
-    def initialize(@hash, @current)
+    def initialize(@hash)
+      @index = 0
```

Comment:
I'd rather track index of a first element (and update it in `delete_entry_and_update_counts`).
If one calls `hash.shift` in a loop, they will suffer from `O(n^2)` without that. (Queue or LRU usecases).

If `@indices_size_pow2` will be changed to `UInt8` and placed near `@indices_bytesize`, then addition if `@first : UInt32` will not change size of hash structure.

Reply:
I'm sorry but I don't understand what you mean.

Reply:
Ah, yes, I see what you mean. I think we should either remove `Hash#shift` or implement your suggestion. We copied `Hash#shift` from Ruby but I don't know what's a real use for that it.

Reply:
Done in https://github.com/crystal-lang/crystal/pull/8017/commits/01e072cdfa948aeb236b642543a7ed89990a346f !

@funny-falcon  I'll be happy if you can review the commit. The change turned out to be pretty simple thanks to the built abstractions.